server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -716,8 +717,20 @@ private DataSegment mergeAndPush(final SegmentIdentifier identifier, final Sink
           closer.register(segmentAndCloseable.rhs);
         }
 
-        mergedFile = indexMerger.mergeQueryableIndex(
-            indexes,
```

Comment:
@kaijianding I agree with @clintropolis, maybe this PR is incomplete, and you planned to free those `indexes`? In it's current form the PR doesn't free any extra resources that were (probably unnecessarily) held before by the worker process.

Reply:
The QueryableIndexes are always new objects loaded from file, and closed when IndexMerger.mergeSegmentFiles() is done. This behavior frees the mmap usage increased during the merge process.
Currently we use indexMerger.mergeQueryableIndex(indexes) in RealtimePlumber, the indexes are not closed to explicitly release the increased mmap because these indexes are still used by queries. Then the mmap usage always increases until abondonSegment() is called which only happens when handoff succeeds.
If handoff is slow or coordinator is not working properly, the mmap usage will keep increasing, this is a problem.

This PR separates the QueryableIndex used by query and used by merge process, the we can close the QueryableIndexes used by merge process to release mmap, and leave the QueryableIndexes used by query untouched.

Reply:
Again, this is beyond me how just creating (and later closing) some new objects without revoking creating _any_ other objects could improve anything, unless any code around here is lazy, that doesn't seem to be the case.

Reply:
If this is about avoiding refreshing some memory mapped files in memory (although I don't see the mechanism how it helps either), at very least the surrounding `try {}` block should be refactored, because currently it couples the creation of the `indexes` list and `mergeSegmentFiles()` for no apparent reason.

Reply:
Will address these comments.

The mechanism is like this:
1. mmap usage will increase after indexes merged. mmap can only be released when close() called on indexes.
2. currently close() on the indexes are not called until abandonSegment() which can be delayed for a very long time if handoff is slow. Before handoff succeeds, the mmap will keep increasing.
3. to avoid this increasement, I create new QueryableIndexes object and close them after each merge success/fail, then the increased mmap during the merge process is released. 
4. I think this magic works because even a flle is mmaped multiple times, the mmap usage is calculated separately each mmap action, and can be un-mmaped separately each un-mmap action. In this case, close() on QueryableIndex is the un-mmap action.

Wish I explain the mechanism clearly. Though this PR is very simple, it indeed help on the mmap usage.

There is an alternative solution to recycle mmap: load the mergedFile as QueryableIndex and swap all hydrants' small QueryableIndexes with this single big QueryableIndex to the Sink. and close all hydrants' small QueryableIndexes and delete all hydrant segments to release mmap (the mmap usage is already increased after merge process done)

Reply:
Ah, I think I am maybe starting to understand what is going on. I was not aware of how the YARN oom killer functions - it turns out, it behaves in a manner I find very unintuitive, where ALL physical memory associated to the process is counted by this killer, _including_ the physical memory occupied by any page cache from our mapped intermediary files that has been tracked by Linux memory accounting as belonging to the process. This is counter to the behavior of the oom killer in the Linux kernel or via cgroups, where the free page cache would be reclaimed by the kernel when memory gets tight before invoking the oom killer. 

See this discussion for details: https://issues.apache.org/jira/browse/HDFS-5957?focusedCommentId=13907875&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13907875 which has spawned this YARN issue https://issues.apache.org/jira/browse/YARN-1747. Is this maybe this issue you are running into? [This solution](https://jira.apache.org/jira/browse/YARN-1775) looks like it has maybe added a setting and been around for a while and might be a possible fix for your problem? If you're running a newer version of YARN, maybe is it possible to use the cgroups solution to limit your YARN processes? I believe that cgroup memory limits will correctly steal from page cache when a process reaches it's memory limits. If that isn't possible, maybe you could disable the physical memory check for containers, or increasing the limits on the container size?

I think the behavior of the YARN physical memory check is sort of nonsensical, at least for the nature of Druid realtime tasks, but maybe you are stuck with it being on as is for whatever circumstance, so back to this PR.

I naively expected mmap from the same process to the same file to be counted as the same memory usage (will look into this further), but if not (as the results of your testing seem to indicate), and accounting is done per call to mmap and knowing that a call to mmap a file itself doesn't actually take physical space until you read data from the file, I could see how this maybe could temporarily trick YARN into not killing your process. This would probably be more likely to hold true if also either: not all columns of all on disk segments had not been queried yet (likely), or if the data was primarily still on heap in incremental indexes and was flushed to do the merge. 

However, to clarify something, I think the mmap usage should not be increasing _after_ the merge, but rather _as a result of the merge_ where all columns of all persisted files from the hydrants are fully read to construct the new segment. Presumably there is free memory available while the merge is happening, that the kernel will allow to be used by page cache for the mapped files since it doesn't know or care about the YARN limits, and that is why this is getting triggered.

I also would think that it means that if queries were to occur after you close the merge only queryable indexes, that ended up covering all columns of all of the segments of the queryable indexes that you _didn't_ close, then you would trigger the same oom killer condition with YARN if there was memory available to be used for page cache, so if anything this strategy would be a trick to avoid the inevitable if data is fully queried. I suspect that depending on the timing of this YARN check, or if the merge were slightly larger, or took longer for any other reason, then the oom killer could still kill a process during merge that would have otherwise been fine, were it not being run with YARN. 

I'm not certain what is the best way to proceed here if you are unable to adjust settings, since I think this fix is for a YARN specific problem but isn't guaranteed to work in all case if what I think is going on is correct. That is because it doesn't actually do anything to the peak memory usage that this realtime task could use if that data were to be fully queried and the OS itself had free space beyond the YARN limits. 

Slow handoff is a bad place for any realtime index tasks to be in, but I can see how this is definitely worse for YARN if it's counting this page cache usage against container limits. The realtime indexing tuning config, `handoffConditionTimeout` is designed to help mitigate this in the normal case, by allowing a realtime indexer to just drop the segments after a period if handoff doesn't happen, but I don't think that would help you in this case, and it has a downside that it can result in data that has been indexed to be unavailable for query if handoff has completely stalled, and I believe it's more oriented to ensure that tasks do not run forever. 

Your proposed alternative fix is potentially interesting, and if the merged segment was smaller than the intermediary files, then it could indeed have a smaller footprint, though it would definitely be a bit more complicated of a fix to ensure everything happens correctly. Ultimately though in your case, I would expect the merged segment would have to be dramatically smaller to avoid the peak usage problem, where if all columns of the merged segment were queried and the OS had the space to cache it then YARN would probably still kill it.

Does this sound correct? (Apologies for such a long response!)

Reply:
Basically correct except for the last part.
The merged segment is not loaded as index after merge process done, so there is no extra mmap used after the merge process is finished, therefore no footprint left.

BUT IF we do load the merged segment as index and replace the hydrant segments in the sink, because the new merged segment is not read thus no mmap used just after the replacement, the hydrant segments definitely use more mmap since there maybe already be queried.
And another benefit of the replacement is that we can reduce the the file number to reduce the inode number. In my case, I'm working one a new scenario, I have lots of hydrant segment and better to reduce inode number after merge.

Reply:
>Basically correct except for the last part.
The merged segment is not loaded as index after merge process done, so there is no extra mmap used after the merge process is finished, therefore no footprint left.

Which part is not correct? Sorry if there was confusion, the hypothetical about what would happen if loading the merged segment after merge was in response to an earlier suggestion, I didn't mean to imply that it is currently loaded. But the hydrant segments that are not closed are loaded still, and querying them after merge will still cause physical memory to be occupied with the page cache.

My main point was that even with this strategy, and regardless of whether or not we keep using the intermediary hydrant segments or swap to the merged segment, any queries _after_ merge will still require physical memory for the cache of reading data out of those memory mapped segments. If the complete set of data is queried (all rows/all columns) then its maximum footprint in terms of physical memory from mmap is the size on disk, if the OS has the free space to give it. In the case of continuing to serve the intermediary segments (as in this PR) this is potentially the exact same footprint as the merge required in the first place, so YARN could still kill it.

Reply:
Cool, glad we could sort out what is going on, and apologies it took so long for me to understand what the point was :+1:

Since this doesn't really seem to particularly have any negative consequences, and since it is doing something useful in some cases at least, I'll have another look at this PR.

Also in jdk 10+, I believe it will be [possible to open files with `O_DIRECT`](https://bugs.openjdk.java.net/browse/JDK-8164900), which is probably what we really want here if we want merge to be done out of the query path and not have a significant impact on page cache usage, though it would potentially lower merge performance. I think it would be worth putting a note and adding this link so our future selves remember to consider if we ever make it out of java 8.

Reply:
@clintropolis it's possible to use `O_DIRECT` already, obtain a raw pointer from `mmap()` and wrap as https://github.com/DataSketches/memory.

Reply:
so my understanding is that, this is useful in case when, on realtime process, very small subset of columns/data in the intermediate persisted segments is being read from queries.
some commentary in the code along those lines would be great.
overall, this patch looks useful to me.

"swap after merge" is further useful to reduce inode count, but the swap would be tricky to implement as there might be queries underway on the indexes just merged.

I think O_DIRECT is whole other beast and would require significant performance regression testing in this case.

Reply:
Hmm, this got marked stale, but i also think it would be ok to merge if @leventov's comments and some additional javadocs explaining what was going on and linking to this discussion were added. Do you have any interest in finishing this @kaijianding?

Apologies that it got stalled in review for so long.